### ISSUE TITLE
Is the model in the final model release path of rppg toolbox the final model？

### ISSUE THREAD

**Issue by jungyin** [zizheng-guo/RhythmFormer#4]
I am trying to use rppg toolbox to try out RhythmFormer, but no matter how I adjust it, I cannot reproduce the effect in the paper. In the field testing process, the accuracy is even worse (only comparing heart rate with PPG, not comparing anything else). May I ask if the models in this directory are the final trained ones? How can I improve my code?

this is my code，and i build some rppg-toolbox code,for my work can do .

```python
#!/usr/bin/env python
# coding=utf8

import scipy

import numpy as np
import cv2
from scipy.signal import welch
from blazeface.blazeface import BlazeFace
import torch
from main import oneRun
import time
import mat73
import matplotlib.pyplot as plt
import threading
import time
import matplotlib
from neural_methods.loss.PhysFormerLossComputer import TorchLossComputer
import heartpy as hp
from scipy.signal import butter
from scipy.signal import resample
from dataset.data_loader.BaseLoader import  BaseLoader
# from unsupervised_methods.methods.OMIT import OMIT


from evaluation.post_process import _detrend,_calculate_SNR,_next_power_of_2,_calculate_fft_hr
matplotlib.use('TkAgg')
import torch
DATA_TYPE = 'Standardized'


def diff_normalize_data(data):
    """Calculate discrete difference in video data along the time-axis and nornamize by its standard deviation."""
    n, h, w, c = data.shape
    diffnormalized_len = n - 1
    diffnormalized_data = np.zeros((diffnormalized_len, h, w, c), dtype=np.float32)
    diffnormalized_data_padding = np.zeros((1, h, w, c), dtype=np.float32)
    replace_value =1e-7
    for j in range(diffnormalized_len):
        dadd =  data[j + 1, :, :, :] + data[j, :, :, :]
        dadd = np.where(dadd == 0, replace_value, dadd)
        diffnormalized_data[j, :, :, :] = (data[j + 1, :, :, :] - data[j, :, :, :]) / (
                dadd )
    diffnormalized_data = diffnormalized_data / np.std(diffnormalized_data)
    diffnormalized_data = np.append(diffnormalized_data, diffnormalized_data_padding, axis=0)
    diffnormalized_data[np.isnan(diffnormalized_data)] = 0
    return diffnormalized_data

def standardized_data(data):
    """Z-score standardization for video data."""
    data = data - np.mean(data)
    data = data / np.std(data)
    data[np.isnan(data)] = 0
    return data

gpu = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
# gpu = torch.device("cpu")
plt.ion()
url = 0

cap = cv2.VideoCapture(url) #设备号为0
cap.set(cv2.CAP_PROP_POS_MSEC, 0)

RhythmFormer = oneRun('configs/infer_configs/UBFC-rPPG_PURE_RHYTHMFORMER.yaml')
RhythmFormer.eval()
model = RhythmFormer

maxframe = 1
norFrame = []

# 准备去计算的frame帧，每30fps为一次推理计算
dframes = []
dframes2 = []

# 识别出的波峰图
detectBVp=[]
# frames1 = diff_normalize_data(frames)
hr_value=0
lastframe = None

ret = True
faces=None

maxfraemlen = 876
maxfraemlen = 160
chunk = 160
widthheight = 128
time123 = time.time()
i1 = 0
hr_value = 0
snrvalues = []
snrvaluedel = 0
fs = 30
lastRppg = []
baseloader = BaseLoader('pure','',None)
while(ret):
    if cap.isOpened() == False:
        print('can not open camera')
        break
    time123=time.time()
    ret, frame = cap.read() #读取图像
    if not ret or frame is None:
        break

    frame = cv2.cvtColor(frame,cv2.COLOR_BGR2RGB)
    

    if(i1>30 ):
        if(len(dframes)<maxfraemlen):
            dframes.append(frame)
        if len(dframes)==maxfraemlen :
            with torch.no_grad():
                dframes = baseloader.crop_face_resize(np.array(dframes),True,'HC',True,1.5,False,30,False,widthheight,widthheight)

                if DATA_TYPE == 'Standardized':
                    dfs = standardized_data(dframes)
                elif DATA_TYPE == 'DiffNormalized' :
                    dfs = diff_normalize_data(dframes)
                dout = np.array(dfs)
            
                dout=np.transpose(dout, (0,3, 1, 2))
                dout=torch.from_numpy(np.array([dout],dtype=np.float32)).cuda()
                rPPG = model(dout)
            def workRppg(rPPG):
                # rPPG = rPPG 
                rPPG = (rPPG-torch.mean(rPPG, axis=-1).view(-1, 1))/torch.std(rPPG, axis=-1).view(-1, 1) 
                rPPG = rPPG.detach().cpu().numpy()
                [b, a] = butter(1, [0.75 / fs * 2, 2.5 / fs * 2], btype='bandpass')
            
                rPPG =rPPG[0]
                if DATA_TYPE == 'Standardized':
                    rPPG =  _detrend(rPPG,100)
                elif DATA_TYPE == 'DiffNormalized' :
                    rPPG =  _detrend(np.cumsum(rPPG),100)
                rPPG1 = scipy.signal.filtfilt(b, a, np.double(rPPG))
                return rPPG1,rPPG
            
            rPPG1,rPPG = workRppg(rPPG)

            hr_value= _calculate_fft_hr(rPPG1,fs)
         
            # dframes =dframes[1:maxfraemlen]
            dframes=[]
            dframes2=[]
       
            print(hr_value)
            plt.clf()
            plt.title('Sine Wave')
            plt.xlabel('Time (s)')
            plt.ylabel('Amplitude')
            plt.grid(True)
            plt.plot(rPPG1,label="rPPG1")
            
            plt.legend()
            
            lastRppg=rPPG1
            t = np.linspace(0, 1, maxfraemlen, endpoint=False)
          
        frame = np.array([frame])
        hr_str = hr_value
    
        if(len(dframes)>0):
            cv2.imshow('frame2',dframes[-1])
        if(faces != None):
            frame =frame 
        else :
            frame = frame
        dt = (time.time()-time123)
      
        mykey = cv2.waitKey(40)
        time123 = time.time() 
        if mykey & 0xFF == ord('q'):
            faces=None
    #         break
    i1+=1
np.save("detectBVp.npy",np.array(detectBVp))
#释放资源
cap.release()
cv2.destroyAllWindows()


```

**Comment by zizheng-guo** [comment]
The RhythmFormer code included in the rPPG-toolbox was not developed by us, so I am not certain whether it is the final model. This issue might help you: https://github.com/zizheng-guo/RhythmFormer/issues/2. I recommend trying the models from this RhythmFormer repository.

Additionally, it seems that you want to infer on your own videos. If the style of the videos you are testing differs significantly from the training set, it may substantially impact performance. The model you are using, trained on the UBFC dataset, might be less stable due to the relatively simple nature of that dataset. You could try models trained on other datasets to see how they perform.

**Comment by jungyin** [comment]
Thank you for your reply. I will conduct another test according to your suggestion
